### ISSUE TITLE
Modify `ChangeDetection.IsKnownImmutableType` to support custom immutable types w/o perf impact

### ISSUE THREAD

**Issue by alexyakunin** [dotnet/aspnetcore#57776]
## Summary

The last line in https://github.com/dotnet/aspnetcore/blob/main/src/Components/Components/src/ChangeDetection.cs#L45 checks for implementation of a very specific `IEventCallback` interface. If it's implemented, the type is treated as immutable.

Let's make `IEventCallback` a descendant of tagging `IBlazorImmutable` (the name is questionable, maybe we need `IImmutable` in .NET?) and use it instead in `IsKnownImmutableType` code.

## Motivation and goals

There are cases where you'd rather detect parameter change with nearly zero false positives vs dealing with more false positives, which typically trigger re-renders or other logic that depends on component parameters. So it's good to have at least one easy option allowing you to declare a parameter type that relies on custom equality.

## In scope / out of scope

Not sure what to write here: assuming we're talking about an interface specific to Blazor only (~ `IBlazorImmutable`), the proposed change is quite isolated.

## Risks / unknowns

I don't see any risks except the scenario when developers implement `IBlazorImmutable` w/o weighting the cost of equality check for this type.

## Examples

```cs
public record UserId : IBlazorImmutable { ... }
public record Avatar : IBlazorImmutable { ... }
public SmallSet<T> : IBlazorImmutable { ... }
```

# Detailed design

```cs
// New interface
public interface IBlazorImmutable;

// Change:
public interface IEventCallback: IBlazorImmutable { ... }

// Change on L45 in ChangeDetection.cs:
// from:
|| typeof(IEventCallback).IsAssignableFrom(type);
// to:
|| typeof(IBlazorImmutable).IsAssignableFrom(type);
```

**Comment by alexyakunin** [comment]
@javiercn sorry, but I can't agree with you:

1. Blaming yourself in allowing others to make mistakes is kinda bad excuse for making something closed rather than open.
2. What's available now is, honestly, a bad option, coz it makes people build way more hacky solutions for problems they still have to tackle.

I certainly know about `SetParametersAsync` - https://github.com/ActualLab/Fusion/blob/master/src/ActualLab.Fusion.Blazor/Components/FusionComponentBase.cs#L23 , and feel free to check out how big is the rabbit hole. There are customizable `ParameterComparer`-s, etc., and all of this is necessary mainly to make sure the state of certain components is recomputed only when it makes sense.

There is way more code than it should be, of course - thanks to, as you've said, "the algorithm that is not extensible on purpose".

Not sure if it's clear, but the fact you make it closed like this doesn't solve any problem - it just creates a bigger one. Overriding `SetParametersAsync` like we already do is one of such ways. But nothing prevents me from "implementing" fake `IEventCallback` as well to make certain parameters compared relying on regular equality. And as soon as this "solution" becomes a well-known one, the whole idea of closing the API like this becomes way more arguable than it seems now.

**Comment by javiercn** [comment]
@alexyakunin thanks for the additional details.

A few things:
* I don't think you should "abuse" IEventCallback in that way. We aren't going to block you but be 100% sure that it's not a supported scenario and that it might break in the future.
* If we wanted to support custom equality, we would probably call `Equals` rather than our own interface.
* I've looked at the implementation you provided. It's unclear to me why you are comparing the properties instead of capturing the values from the `ParameterView` into a dictionary and comparing them against the new ones.

We might disagree on goals here:
* Triggering an extra render might introduce a performance penalty when the values have indeed not changed.
* Not triggering a render will translate in an incorrect behavior of the app, that is also, really hard to debug.

In this case, we favor correctness over potential performance gains because we believe is much worse to run into a situation where you have a hard to debug issue than the cost of potentially triggering a few additional renders.

If you have components where rendering is expensive you can control when they render via `ShouldRender` or overriding `SetParametersAsync`. That's the way we recommend people handle this situation, as opposed to interfering with the diffing algorithm that has a much broader impact.

**Comment by alexyakunin** [comment]
@javiercn lol, I didn't notice `IEventCallback` is internal - i.e. yeah, it's hard to abuse it that way.

> If we wanted to support custom equality, we would probably call Equals rather than our own interface.

IMO it's totally reasonable to make sure that default equality isn't used here:
- Yes, most of data structures aren't immutable, so it just doesn't make any sense to use it
- And yes, `Equals` can be expensive.

So IMO it's the right call to use "more pessimistic" equality here.

But notice that my suggestion doesn't break any benefits of the existing solution:
- It's reasonable to assume that if you tag a type with `IBlazorImmutable`, you know the consequences of this.
- This action is very explicit - i.e. you can't do this unintentionally
- The computational cost is also the same.

**Comment by alexyakunin** [comment]
> * I've looked at the implementation you provided. It's unclear to me why you are comparing the properties instead of capturing the values from the `ParameterView` into a dictionary and comparing them against the new ones.

All depends on the number of such parameters you actually have: if it's about just a few components & params, of course you can manually copy the old value for each param & compare it with the new one in `SetParametersAsync` to decide whether the change had happened.

But what if it's about nearly every parameter in `ComputedStateComponent<T>`? There can be hundreds of such components, each with a few params they use in `ComputeState(...)`. So when these params change, the state has to be recomputed, which typically triggers re-render.

You can certainly try to manually copy & compare all of such params, but this is definitely more error prone than using a solution which may address the problem as a whole.

**Comment by alexyakunin** [comment]
> We might disagree on goals here:
> 
> * Triggering an extra render might introduce a performance penalty when the values have indeed not changed.
> * Not triggering a render will translate in an incorrect behavior of the app, that is also, really hard to debug.

Yes, I fully understand this. Just want to point out that *suggested change doesn't break these assumptions*.

> If you have components where rendering is expensive you can control when they render via `ShouldRender` or overriding `SetParametersAsync`. That's the way we recommend people handle this situation, as opposed to interfering with the diffing algorithm that has a much broader impact.

That's exactly what we do. You may notice there is e.g. https://github.com/ActualLab/Fusion/blob/master/src/ActualLab.Fusion.Blazor/Components/ComputedRenderStateComponent.cs , which cancels renders unless the state it computes gets changed. It's not super straightforward from the code, but ultimately, it assumes that component's `State` (it's what `ComputedStateComponent<T>` computes/updates when `State` dependencies change) is the only data source for its render logic. So unless it changes, it doesn't have to re-render anything. And it uses `State.Snapshot` to track the changes.

Long story short, we know how it works pretty well.

**Comment by alexyakunin** [comment]
I also want to highlight that you may want to come up with a better solution - i.e. IMO what I suggested is also sub-optimal:
- It allows to override the "blazor equality" only for your own types
- So if you want to override it for something else, it won't work

And IMO this alone is enough to reject this proposal as "not good enough". But let's come up with what's good enough in this case - I hope it's clear that having an ability to override "blazor equality" is important, and personally I'm fine with any good option here (e.g. making `IChangeDetection` DI-injectable, etc.).

**Comment by javiercn** [comment]
@alexyakunin thanks for the additional details.

I think we disagree at which level this should happen. For us:
* The decision on when to render belongs to the component itself, not outside of it.
* The change detection we perform during diffing is an optimization to avoid setting parameters in common cases where we know they haven't changed. It's meant to be deterministic and simple, and favor correctness over anything else.

Some more details:
    * If we were to provide an `IComponentParametersComparer` service that you could register on DI to provide your own implementation, we would open the door for anyone changing when/how third-party components render.
      * This would mean that component vendors and authors would have to account for these situations in their component's implementation.
    * If we were to provide an interface that you could implement on a type, you could not change existing types, but you could still affect the rendering characteristics of generic components as the diffing would be influenced by the implementation on the given generic value.

Rendering is a very sensitive area of the codebase and a hot path, having other code running there creates all sort of opportunities to issues, from issues like "Why is my component not re-rendering" to "Why is my app suddenly slow" that can happen in very non-obvious ways.
* You use a type with a bad/slow implementation.
* You bring in your own change detection logic with bugs/performance issues.

The current approach offers predictability and stability both in terms of performance and when the component receives parameters. Any decision on when the component renders need to be performed at the component level itself, not on a central place that can have side-effects on the way other components receive parameters.

If we were to do something in this area, we would simplify the process of implementing this on a per-component level, but we wouldn't change this at the diffing level.

To follow with the example, you provided and as an alternative approach, source generation could be used to generate a method to perform the comparisons. Note that the implementation can be potentially simpler than the current one you have. You'd need to compare only parameters that aren't supported out of the box, not all the parameters, because if all supported parameters are equal, the component wouldn't receive a new set.

**Comment by alexyakunin** [comment]
Based on what I just read, my original suggestion w/ `IBlazorImmutable` fully fits the bill:
- Unless you're a developer authoring the component, you can't influence on how the comparison works for it - because its parameter types are fixed (ok, unless some of them are generic, which is extremely rare case).
- There is no way to change the comparison logic on global level.

In other words, I don't understand your argument. Its applicable for another option I mentioned (extracting `IChangeDetection`), but not for the original one.

And... I'm not sure if you noticed, but I very explicitly saying that you advocate nearly this policy:

```txt
Developers: 
- Hey, why you give us a hand saw? We need a chainsaw!
Microsoft:
- Sorry, I can't give it to you, because you can cut your fingers.
Developers:
- But we need to cut through the forest! And it's our fingers, not yours - let us decide whether to risk them or not, ok?
Microsoft:
- Sorry, I can't see the forest in the toy apps and demos I've built for you.
Developers:
- Well, your 10K LOC demos aren't our 1M LOC codebases. That's why you don't see a plethora of other problems too.
Microsoft:
- I still insist my decision is correct. Also notice that I let you use a knife in addition to hand saw.
...
```

And speaking of source generators:

> source generation could be used to generate a method to perform the comparisons

Even if we put aside all the pain and suffering you need to go through to implement a single one (believe me, I know what I'm talking about), **I would be delighted to see A SINGLE source generator producing the source based on .razor / Blazor component source.** 

AFAIK there is a plethora of problems with this:
- AFAIK source generators can't consume the output of other source generators.
- Even the ordering is a problem: https://github.com/dotnet/roslyn/issues/57239
... And tons of other issues you can easily Google.

**Comment by alexyakunin** [comment]
> We don't believe excessive re-renders because of parameter comparisons represent a problem that spreads across many components in a given app.

I love your level of confidence :)

Ok, maybe it makes sense to point to one obvious case: typed identifiers.
- Record structs are ~ zero-cost abstraction allowing you to wrap e.g. strings to `UserId`. And the later are much more convenient to use - e.g. you won't pass `UserId` to a method expecting `OrderId`.
- And if you have these types, it's natural to use them as component parameters, right?

Now ask a random Blazor developer about the effect of trying to do so - at scale.

**Comment by alexyakunin** [comment]
> Libraries can still provide types that implement `IBlazorImmutable` that when used on components will have an impact.

YES, AND THAT'S EXACTLY THE REASON TO ADD IT! 

I feel like I talk with ChatGPT here. Or maybe there is an unspoken rule in Microsoft, which goes like this: "If you're brave enough to suggest some change, make sure it changes absolutely nothing. Otherwise it's no-no-no."

It's amazing to see that at the very same company one group of people work on language with a decently complex type system  that brings type safety & lots of other benefits, if you use it right. And there is another group, who literally says, "We're absolutely sure the dozen of primitive types we pre-selected covers every single case you can dream of. Oh, wait - we almost forgot to add `Guid` there!" , and all of this is justified by... Presumed stupidity of an average developer, right?

Sorry, but "640KB should be enough for everyone" is way more logical than this.

**Comment by alexyakunin** [comment]
The more I think about this, the worse it becomes.

Did you guys ever dream of making `Equals` and `GetHashCode` sealed? Please be honest.

**Comment by matthetherington** [comment]
There have been many discussions on here around changing Blazor's default change detection - and I personally think *some change* there is warranted - but I haven't even seen agreement that there's a problem to address so it seems unlikely to change anytime soon.

Manually writing parameter change detection in each component across an entire app, something that intuitively you would expect to happen out of the box, can get tedious quickly. And `ShouldRender` is awkward for some of our needs, as:

1. `ShouldRender` is part of the component, and for any given component, you can't easily have different logic for deciding when to re-render in different places where the component is used.
2. Sometimes you don't control `ShouldRender` - eg maybe you're using a component library and the type or method is sealed.
3. Without also overriding SetParametersAsync and diffing parameters, data-fetching code inside OnParametersSetAsync still gets ran, leading to redundant network requests, database queries, etc.

It would be very helpful if there was a way that eg records or DDD-style value objects can be detected as unchanged across parameter changes in addition to the existing primitive types. Having only a small set of types behave intuitively in this way is quite surprising.

We use custom `Identifier` classes, for example, instead of passing around primitives, so instead of:

```
[Parameter] public int UserId { get; set; }
```

where rendering would be skipped if `UserId` was unchanged, and the component had no other complex parameters, we have:

```
[Parameter] public UserIdentifier UserId { get; set; }
``` 

which causes the component to re-render no matter what. It's frustrating that Blazor penalises us for something we consider to be good practice outside of Blazor (avoiding primitive obsession, that is). If there is good justification for the existing implementation, then perhaps a mechanism to opt-in to `Equals` based change detection on a per component basis via an attribute (like `@rendermode`) would work?

In the meantime, I put together a `Memo` component for Blazor - sort of like React's `useMemo` - which will compare a provided `Keys` parameter in the previous and incoming parameters, and only re-render if a key value changed:

```cshtml
@using BlazorMemoire.Components

@* ExpensiveChildComponent will only re-render when user.Id or selectedTab changes. *@
@* Changing from one user instance to another with the same ID will mean ExpensiveChildComponent doesn't re-render *@
@* Blazor's change detection would always re-render ExpensiveChildComponent when the parent changes, as `User` is a complex type *@

<Memo Keys="@([user.Id, selectedTab])"> 
    <ExpensiveChildComponent User="user" Tab="selectedTab" />
</Memo>
```

It’s a shallow-compare by default (via `object.Equals`) with an option for a more-involved deep comparison which allows comparing `List<T>` parameter elements, `HashSet<T>`, `IEnumerable`, etc too.

It's published to Nuget as `BlazorMemoire` if anyone wants to give it a try. The repo's over at: https://github.com/matthetherington/Blazor-Memoire - any feedback you might have is very welcome.
